pkg/utils/etcdutil/etcdutil_test.go

```diff
@@ -252,45 +230,145 @@ func TestEtcdClientSync(t *testing.T) {
 	cfg2.ClusterState = embed.ClusterStateFlagExisting
 	peerURL := cfg2.LPUrls[0].String()
-	addResp, err := AddEtcdMember(client1, []string{peerURL})
+	addResp, err := AddEtcdMember(client, []string{peerURL})
 	re.NoError(err)
 	etcd2, err := embed.StartEtcd(cfg2)
-	defer func() {
-		etcd2.Close()
-	}()
 	re.NoError(err)
 	re.Equal(uint64(etcd2.Server.ID()), addResp.Member.ID)
 	<-etcd2.Server.ReadyNotify()
+	return etcd2
+}
 
+func checkMembers(re *require.Assertions, client *clientv3.Client, etcds []*embed.Etcd) {
 	// Check the client can get the new member.
-	listResp2, err := ListEtcdMembers(client1)
-	re.NoError(err)
-	re.Len(listResp2.Members, 2)
-	for _, m := range listResp2.Members {
-		switch m.ID {
-		case uint64(etcd1.Server.ID()):
-		case uint64(etcd2.Server.ID()):
-		default:
-			t.Fatalf("unknown member: %v", m)
+	listResp, err := ListEtcdMembers(client)
+	re.NoError(err)
+	re.Len(listResp.Members, len(etcds))
+	inList := func(m *etcdserverpb.Member) bool {
+		for _, etcd := range etcds {
+			if m.ID == uint64(etcd.Server.ID()) {
+				return true
+			}
 		}
+		return false
 	}
+	for _, m := range listResp.Members {
+		re.True(inList(m))
+	}
+}
 
-	// Remove the first member and close the etcd1.
-	_, err = RemoveEtcdMember(client1, uint64(etcd1.Server.ID()))
+func proxyWithDiscard(re *require.Assertions, server, proxy string, enableDiscard *atomic.Bool) {
+	server = strings.TrimPrefix(server, "http://")
+	proxy = strings.TrimPrefix(proxy, "http://")
+	l, err := net.Listen("tcp", proxy)
 	re.NoError(err)
-	time.Sleep(20 * time.Millisecond) // wait for etcd client sync endpoints and client will be connected to etcd2
-	etcd1.Close()
+	for {
+		connect, err := l.Accept()
+		re.NoError(err)
+		go func(connect net.Conn) {
+			serverConnect, err := net.Dial("tcp", server)
+			re.NoError(err)
+			pipe(connect, serverConnect, enableDiscard)
+		}(connect)
+	}
+}
 
-	// Check the client can get the new member with the new endpoints.
-	listResp3, err := ListEtcdMembers(client1)
-	re.NoError(err)
-	re.Len(listResp3.Members, 1)
-	re.Equal(uint64(etcd2.Server.ID()), listResp3.Members[0].ID)
+func pipe(src net.Conn, dst net.Conn, enableDiscard *atomic.Bool) {
+	errChan := make(chan error, 1)
+	go func() {
+		err := ioCopy(src, dst, enableDiscard)
+		errChan <- err
+	}()
+	go func() {
+		err := ioCopy(dst, src, enableDiscard)
+		errChan <- err
+	}()
+	<-errChan
+	dst.Close()
+	src.Close()
+}
 
-	require.NoError(t, failpoint.Disable("github.com/tikv/pd/pkg/utils/etcdutil/autoSyncInterval"))
+func ioCopy(dst io.Writer, src io.Reader, enableDiscard *atomic.Bool) (err error) {
+	buffer := make([]byte, 32*1024)
+	for {
+		if enableDiscard.Load() {
+			io.Copy(io.Discard, src)
+		}
+		readNum, errRead := src.Read(buffer)
+		if readNum > 0 {
+			writeNum, errWrite := dst.Write(buffer[0:readNum])
+			if errWrite != nil {
+				return errWrite
+			}
+			if readNum != writeNum {
+				return io.ErrShortWrite
+			}
+		}
+		if errRead != nil {
```

Comment:
if errRead == EOF, should we return EOF or return nil to the caller of this function?

Reply:
I think direct return is enough.

Reply:
Maybe return nil